File: src/BattalionHQ.cpp
#include "BattalionHQ.h"
#include <cstdio>

BattalionHQ::BattalionHQ()
    : designation_("MINDER-HQ")
    , status_(HQStatus::OPERATIONAL)
    , isRelocating_(false)
    , relocateTimeRemaining_(0.0f)
    , radarOnline_(false)
    , commsOnline_(false)
{
    position_.range = 0.0f;
    position_.azimuth = 0.0f;
    relocateDestination_.range = 0.0f;
    relocateDestination_.azimuth = 0.0f;
}

BattalionHQ::~BattalionHQ() {}

void BattalionHQ::init(float posRange, float posAzimuth)
{
    position_.range = posRange;
    position_.azimuth = posAzimuth;
    status_ = HQStatus::OPERATIONAL;
    isRelocating_ = false;
    relocateTimeRemaining_ = 0.0f;
    radarOnline_ = true;
    commsOnline_ = true;
}

void BattalionHQ::update(float dt)
{
    if (!isRelocating_) return;

    relocateTimeRemaining_ -= dt;

    if (relocateTimeRemaining_ <= 0.0f) {
        // Relocation complete — fully operational at new position
        position_ = relocateDestination_;
        isRelocating_ = false;
        relocateTimeRemaining_ = 0.0f;
        status_ = HQStatus::OPERATIONAL;
        radarOnline_ = true;
        commsOnline_ = true;
    } else if (relocateTimeRemaining_ <= SETUP_TIME) {
        // Setting up at new position — radar coming online first
        status_ = HQStatus::SETTING_UP;
        // Radar comes online in last 15 seconds of setup
        radarOnline_ = (relocateTimeRemaining_ <= SETUP_TIME * 0.5f);
        commsOnline_ = false;
    } else {
        // Still moving or tearing down
        status_ = HQStatus::RELOCATING;
        radarOnline_ = false;
        commsOnline_ = false;
    }
}

bool BattalionHQ::relocate(float newRange, float newAzimuth)
{
    if (isRelocating_) return false;
    if (status_ == HQStatus::SETTING_UP) return false;

    relocateDestination_.range = newRange;
    relocateDestination_.azimuth = newAzimuth;
    isRelocating_ = true;
    relocateTimeRemaining_ = TOTAL_RELOCATE_TIME;
    status_ = HQStatus::RELOCATING;
    radarOnline_ = false;
    commsOnline_ = false;

    return true;
}

void BattalionHQ::cancelRelocation()
{
    if (!isRelocating_) return;

    // Emergency halt — stay at current position, bring systems back online
    isRelocating_ = false;
    relocateTimeRemaining_ = 0.0f;

    // Need setup time to get operational again at current position
    status_ = HQStatus::SETTING_UP;
    relocateTimeRemaining_ = SETUP_TIME;
    isRelocating_ = true;  // Reuse relocation countdown for setup
    relocateDestination_ = position_;  // Stay in place
}

std::string BattalionHQ::getStatusString() const
{
    switch (status_) {
        case HQStatus::OPERATIONAL: return "OPERATIONAL";
        case HQStatus::RELOCATING:  return "RELOCATING";
        case HQStatus::SETTING_UP:  return "SETTING UP";
        case HQStatus::DEGRADED:    return "DEGRADED";
    }
    return "UNKNOWN";
}

HQData BattalionHQ::getData() const
{
    HQData data;
    data.designation = designation_;
    data.status = status_;
    data.position = position_;
    data.relocateDestination = relocateDestination_;
    data.isRelocating = isRelocating_;
    data.relocateTimeRemaining = relocateTimeRemaining_;
    data.totalRelocateTime = TOTAL_RELOCATE_TIME;
    data.personnelCount = HQ_PERSONNEL;
    data.radarOnline = radarOnline_;
    data.commsOnline = commsOnline_;
    data.statusString = getStatusString();
    return data;
}

std::string HQData::format() const
{
    char buf[512];
    snprintf(buf, sizeof(buf),
        "=== BATTALION HQ ===\n"
        "  UNIT:   %s\n"
        "  STATUS: %s\n"
        "  POS:    %.1f km / %03d deg\n"
        "  RADAR:  %s\n"
        "  COMMS:  %s\n"
        "  STAFF:  %d personnel\n",
        designation.c_str(),
        statusString.c_str(),
        position.range, (int)position.azimuth,
        radarOnline ? "ONLINE (AN/TPS-43E)" : "OFFLINE",
        commsOnline ? "ONLINE (encrypted)" : "OFFLINE",
        personnelCount);

    std::string result(buf);

    if (isRelocating) {
        char reloc[128];
        snprintf(reloc, sizeof(reloc),
            "  DEST:   %.1f km / %03d deg\n"
            "  ETA:    %.0fs remaining\n",
            relocateDestination.range, (int)relocateDestination.azimuth,
            relocateTimeRemaining);
        result += reloc;
    }

    return result;
}

File: cocos-creator-project/assets/scripts/game/BattalionHQ.ts
import { PolarCoord } from './GameTypes';

// ============================================================================
// Battalion HQ — Mobile AN/TSQ-73 Missile Minder Battery
// ============================================================================

export enum HQStatus {
    OPERATIONAL,
    RELOCATING,
    SETTING_UP,
    DEGRADED,
}

export interface HQData {
    designation: string;
    status: HQStatus;
    position: PolarCoord;
    relocateDestination: PolarCoord;
    isRelocating: boolean;
    relocateTimeRemaining: number;
    totalRelocateTime: number;
    personnelCount: number;
    radarOnline: boolean;
    commsOnline: boolean;
    statusString: string;
}

export class BattalionHQ {
    // Real-world relocation timing (in game seconds, representing minutes)
    // Actual AN/TSQ-73 battalion displacement: ~45 min teardown, travel,
    // ~45 min setup. Total 90+ minutes plus convoy travel time.
    // Game uses 1 game-second = 1 real-minute time compression for relocation.
    static readonly TEARDOWN_TIME = 45.0;   // 45 minutes — power down, cables, antenna
    static readonly MOVE_TIME = 30.0;        // 30 minutes — convoy travel (variable)
    static readonly SETUP_TIME = 45.0;       // 45 minutes — erect antenna, cables, boot L-112s
    static readonly TOTAL_RELOCATE_TIME = 120.0; // ~2 hours total displacement
    static readonly HQ_PERSONNEL = 35;

    private designation = 'MINDER-HQ';
    private status: HQStatus = HQStatus.OPERATIONAL;
    private position: PolarCoord = { range: 0, azimuth: 0 };
    private relocateDestination: PolarCoord = { range: 0, azimuth: 0 };
    private _isRelocating = false;
    private relocateTimeRemaining = 0;
    private _radarOnline = false;
    private _commsOnline = false;

    init(posRange: number, posAzimuth: number): void {
        this.position = { range: posRange, azimuth: posAzimuth };
        this.status = HQStatus.OPERATIONAL;
        this._isRelocating = false;
        this.relocateTimeRemaining = 0;
        this._radarOnline = true;
        this._commsOnline = true;
    }

    update(dt: number): void {
        if (!this._isRelocating) return;

        this.relocateTimeRemaining -= dt;

        if (this.relocateTimeRemaining <= 0) {
            this.position = { ...this.relocateDestination };
            this._isRelocating = false;
            this.relocateTimeRemaining = 0;
            this.status = HQStatus.OPERATIONAL;
            this._radarOnline = true;
            this._commsOnline = true;
        } else if (this.relocateTimeRemaining <= BattalionHQ.SETUP_TIME) {
            this.status = HQStatus.SETTING_UP;
            this._radarOnline = this.relocateTimeRemaining <= BattalionHQ.SETUP_TIME * 0.5;
            this._commsOnline = false;
        } else {
            this.status = HQStatus.RELOCATING;
            this._radarOnline = false;
            this._commsOnline = false;
        }
    }

    relocate(newRange: number, newAzimuth: number): boolean {
        if (this._isRelocating) return false;
        if (this.status === HQStatus.SETTING_UP) return false;

        this.relocateDestination = { range: newRange, azimuth: newAzimuth };
        this._isRelocating = true;
        this.relocateTimeRemaining = BattalionHQ.TOTAL_RELOCATE_TIME;
        this.status = HQStatus.RELOCATING;
        this._radarOnline = false;
        this._commsOnline = false;
        return true;
    }

    cancelRelocation(): void {
        if (!this._isRelocating) return;

        this._isRelocating = false;
        this.relocateTimeRemaining = 0;
        this.status = HQStatus.SETTING_UP;
        this.relocateTimeRemaining = BattalionHQ.SETUP_TIME;
        this._isRelocating = true;
        this.relocateDestination = { ...this.position };
    }

    // Getters
    getStatus(): HQStatus { return this.status; }
    isOperational(): boolean { return this.status === HQStatus.OPERATIONAL; }
    isRelocating(): boolean { return this._isRelocating; }
    isRadarOnline(): boolean { return this._radarOnline; }
    isCommsOnline(): boolean { return this._commsOnline; }
    getPosition(): PolarCoord { return { ...this.position }; }
    getRelocateTimeRemaining(): number { return this.relocateTimeRemaining; }

    getStatusString(): string {
        switch (this.status) {
            case HQStatus.OPERATIONAL: return 'OPERATIONAL';
            case HQStatus.RELOCATING:  return 'RELOCATING';
            case HQStatus.SETTING_UP:  return 'SETTING UP';
            case HQStatus.DEGRADED:    return 'DEGRADED';
        }
    }

    getData(): HQData {
        return {
            designation: this.designation,
            status: this.status,
            position: { ...this.position },
            relocateDestination: { ...this.relocateDestination },
            isRelocating: this._isRelocating,
            relocateTimeRemaining: this.relocateTimeRemaining,
            totalRelocateTime: BattalionHQ.TOTAL_RELOCATE_TIME,
            personnelCount: BattalionHQ.HQ_PERSONNEL,
            radarOnline: this._radarOnline,
            commsOnline: this._commsOnline,
            statusString: this.getStatusString(),
        };
    }
}
